xla/pjrt/pjrt_stream_executor_client.cc

```diff
@@ -3234,9 +3235,57 @@ PjRtStreamExecutorLoadedExecutable::GetHloModules() const {
   return std::move(modules);
 }
 
+namespace {
+
+absl::StatusOr<absl::string_view> MemoryKindFromSimpleShape(
+    const Shape& shape, absl::string_view default_memory_kind) {
+  if (!shape.has_layout()) {
+    return default_memory_kind;
+  }
+  switch (shape.layout().memory_space()) {
+    case Layout::kHostMemorySpace:
+      return PinnedHostMemorySpace::kKind;
+    case Layout::kDefaultMemorySpace:
+      return default_memory_kind;
+    default:
+      return InvalidArgument("Unexpected memory space %d in output layout",
+                             shape.layout().memory_space());
+  }
+}
+
+absl::StatusOr<std::vector<absl::string_view>> MemoryKindsFromShape(
+    const Shape& shape, absl::string_view default_memory_kind) {
+  std::vector<absl::string_view> result;
+  if (shape.IsTuple()) {
+    for (auto element_shape : shape.tuple_shapes()) {
+      TF_ASSIGN_OR_RETURN(
+          absl::string_view element_memory_kind,
+          MemoryKindFromSimpleShape(element_shape, default_memory_kind));
+      result.push_back(element_memory_kind);
+    }
+  } else {
+    TF_ASSIGN_OR_RETURN(absl::string_view memory_kind,
+                        MemoryKindFromSimpleShape(shape, default_memory_kind));
+    result.push_back(memory_kind);
+  }
+  return result;
+}
+
+}  // namespace
+
 absl::StatusOr<std::vector<std::vector<absl::string_view>>>
 PjRtStreamExecutorLoadedExecutable::GetOutputMemoryKinds() const {
-  return Unimplemented("GetOutputMemoryKinds is not supported.");
+  TF_ASSIGN_OR_RETURN(auto shapes, GetOutputShapes());
+  TF_ASSIGN_OR_RETURN(PjRtMemorySpace * default_memory_space,
+                      addressable_devices()[0]->default_memory_space());
+  std::vector<std::vector<absl::string_view>> out;
```

Comment:
Done, thank you!

I have also rearranged the MemoryKindsFromShape function so that it constructs the one element vector directly, rather than by pushing.